docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,325 @@
+  kernelspec:
+    display_name: Python 3
+    name: python3
+---
+
+(pallas_mgpu_pipelining)=
+<!-- #region id="bJ5yuIr-M0x0" -->
+
+## Mosaic GPU Pipelining
+
+This guide covers software pipelining using the Mosaic GPU backend for Pallas.
+
+For a general overview of the pipelining API in Pallas, we recommend that users first read {ref}`pallas_software_pipelining`. Pipelining in Pallas is programmed explicitly. For those who are familiar with Triton, this is a significant difference in programming model because in Triton, pipelining is an optimization that is done automatically by the compiler.
+
+<!-- #endregion -->
+
+```python id="dGAa3iO5DoRT"
+import jax
+from jax import lax
+from jax import numpy as jnp
+from jax.experimental.pallas import mosaic_gpu as plgpu
+from jax.experimental import pallas as pl
+import numpy as np
+```
+
+<!-- #region id="Pv9j90hVyswo" -->
+
+### Pipelining with Mosaic GPU
+
+The recommended approach to pipeline using Mosaic GPU is to use the {func}`jax.experimental.pallas.mosaic_gpu.emit_pipeline` function to pipeline over sequential loops (and to use `plgpu.kernel` to partition the problem in parallel over the CUDA grid). `emit_pipeline` follows a similar API as `pl.pallas_call` except it exposes a few additional GPU-specific options.
+
+- `body`, `grid` have similar semantics as in `pl.pallas_call`. The `grid` denotes how many invocations of the `body` function to run. In contrast with a CUDA grid, the pipeline grid is guaranteed to run sequentially.
+- `in_specs` and `out_specs` also similarly to `pl.pallas_call`, except they also accept `plgpu.GPUBlockSpec` instances that can be used specify GPU-specific transforms, such as swizzling. See [memory reference transforms](https://docs.jax.dev/en/latest/pallas/gpu/reference.html#memory-reference-transforms) for more detail on available transformations.
+- `max_concurrent_steps` controls the maximum number of concurrent memory transfers. Using additional concurrent steps will consume more SMEM to hold temporary buffers, so this option should be used carefully.
+- `delay_release` allows the user to specify an additional number of iterations to wait before the buffer is re-used by the pipeline. For example, a buffer copied into SMEM on iteration 0 with `delay_release=1` and `max_concurrent_steps=2` will not be re-used until iteration 3, as opposed to iteration 2 for a standard double-buffered strategy. This is useful for certain optimizations such as allowing multiple async matmuls in flight to keep the tensor core pipeline filled.
+
+#### Compatibility API using `pl.pallas_call`
+
+As an alternative to `emit_pipeline` and to maintain compatibility with Pallas TPU, Mosaic GPU also implements the existing `pl.pallas_call` API for pipelining. Pipelining with `pl.pallas_call` directly requires the user to pass in a `plgpu.GPUCompilerParams` object as the `compiler_params` argument, which specifies the following options that are relevant for pipelining:
```

Comment:
This could be explained better. It is true that no pipelining happens when you don't specify the compiler params, because by default we schedule all grid elements in parallel. You can opt in into pipelining them inside the kernel by specifying them as sequential. This also lets you reuse `scratch_buffers` across iterations, which lets you e.g. partition the reduction dimension of a matmul across the Pallas grid without splitting it across blocks (that would then require cross-block reductions like in split-k).

Reply:
Done.